Image picker plugin: use the native Android gallery / file chooser

roughike opened the pull request:
Changes: 
* Uses the native Android gallery UI for picking images, instead of a custom one
* removes the image source selection dialog on iOS
* **Breaking API change:** now the `source` named parameter for the `pickImage` method is annotated with `@required`.

Issues fixed:
* Fixes flutter/flutter#13706
* Fixes flutter/flutter#11577, except for step 4. Which could be a separate ticket.
* Fixes flutter/flutter#13060, which is an old leftover issue that was actually merged a while ago, but verified working here

Currently, the image picker plugin uses a custom gallery UI from [this library](https://github.com/esafirm/android-image-picker), instead of the native one. I'll guess that the original reason for that is that picking images by using the native gallery can be quite wonky.

While this worked fine on my emulator and Motorola Z, there might be a risk that this breaks picking images from gallery for some devices.

Here's what I mean by "native Android gallery":

![imgpick](https://user-images.githubusercontent.com/13744304/37431554-3d3a87e8-27de-11e8-8747-4d20a0070cee.gif)

I also refactored the image resizing and Exif data copying logic to separate classes to make the code cleaner.

Here's what's currently missing to land this, feel free to give further feedback:

- [x] figure out the license headers: needs input from @Hixie 
- [x] inform the flutter-dev mailing list about the breaking API change: https://groups.google.com/forum/#!topic/flutter-dev/YE5tM2Y0BP8

mtthsfrdrch commented:
hehe, started to do this as well, yesterday, but you were faster :wink: ... the library dependencies on the Android side could be removed as well now :+1:

roughike commented:
Sorry about that 👹

Actually, it still depends on the third-party library to take images with the camera. The plan is to refactor that out too when/if this gets merged first!

Hixie commented on packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/FileUtils.java:
```diff
@@ -0,0 +1,132 @@
+import android.text.TextUtils;
+
+/*
+ * Copyright (C) 2007-2008 OpenIntents.org
```
Move this comment to the top of the file.

Hixie commented on packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/FileUtils.java:
```diff
@@ -0,0 +1,132 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * A modified version of
```
Rather than just "A modified version of", please say something like "This file was modified by the Flutter authors from the following original file:".

Hixie commented:
Please make sure you append https://raw.githubusercontent.com/iPaulPro/aFileChooser/master/LICENSE.txt to the end of https://github.com/flutter/plugins/blob/master/packages/image_picker/LICENSE, separated from the previous license by a single line of exactly eighty `-` characters (hyphens).

goderbauer commented on packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerPlugin.java:
```diff
@@ -104,7 +99,7 @@ public void onMethodCall(MethodCall call, Result result) {
+                REQUEST_CAMERA_PERMISSIONS);
             break;
           }
           activity.startActivityForResult(
```
Add a TODO here that the Camera should also be switched to use the native camera?

jonahwilliams commented on packages/image_picker/lib/image_picker.dart:
```diff
@@ -6,31 +6,20 @@ import 'dart:async';
-  /// On iOS, the user is presented with an alert box with options to
-  /// either take a photo using the device camera or pick an image from
-  /// the photo library.
-  askUser,
```
Removing this feature (and thus the default value) is a breaking change - I think it would be a good idea to send an email to flutter-dev@googlegroups.com letting people know.

roughike replied:
Originally there was no way of choosing a source for the images. The native sides of this plugin had the "source chooser dialogs" on iOS and Android. I introduced an option of controlling the image source in #271, and I think I should've removed the dialogs back then, but I didn't want to break anything.

(that was just me rationalizing why I did this breaking change :))

I can do the email when/if this PR gets approved, thanks for the idea!

Hixie replied:
We prefer to warn people in advance of landing the PRs -- see https://flutter.io/design-principles/#handling-breaking-changes

roughike replied:
Right. I just read it through and that makes a lot of sense, sorry for my ignorance.

I _think_ I just sent the email, but can't see it anywhere. Can you verify it actually went through?

goderbauer commented:
You'll also need add an entry to the CHANGELOG and bump the version in the pubspec.yaml.

roughike commented:
@goderbauer Any input on the new version number? It's a breaking change, so I was thinking of `0.4.0`.

goderbauer commented on packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/FileUtils.java:
```diff
@@ -0,0 +1,137 @@
+/*
```
This differs from the license header in the source file you mentioned below?

roughike replied:
The original authors LICENSE file states `Copyright 2011 - 2013 Paul Burke` and the file header `Copyright (C) 2007-2008 OpenIntents.org`.

I wanted to have this uniformed, but I'm not sure which one to use ¯\_(ツ)_/¯

goderbauer replied:
/cc @Hixie, see above.

Hixie replied:
The file header should be copied verbatim and unchanged.

goderbauer commented on packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/FileUtils.java:
```diff
@@ -0,0 +1,137 @@
+ * https://raw.githubusercontent.com/iPaulPro/aFileChooser/master/aFileChooser/src/com/ipaulpro/afilechooser/utils/FileUtils.java
+ */
+
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
You probably only want one license header (the one from the original file)

/cc @Hixie

Hixie replied:
Yeah, don't add a new license here, this file is under the original license.

goderbauer commented on packages/image_picker/LICENSE:
```diff
@@ -24,3 +24,17 @@ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+--------------------------------------------------------------------------------
+Copyright 2011 - 2013 Paul Burke
```
This is *not* the content of https://raw.githubusercontent.com/iPaulPro/aFileChooser/master/LICENSE.txt?

/cc @Hixie

roughike replied:
[See my previous comment for the reasoning about this.](https://github.com/flutter/plugins/pull/423#discussion_r175929999)

Hixie replied:
This must be the contents of <https://raw.githubusercontent.com/iPaulPro/aFileChooser/master/LICENSE.txt>.

goderbauer commented:
@roughike Yeah, bumping it to 0.0.4 for the breaking change sounds good.

roughike commented:
It's sleepy time in Finland right now. I'll get back to this tomorrow.

At least the license & file headers have to be fixed. Also, the example in the README must be modified a bit; the current one won't compile because of the breaking API changes.

mit-mit commented:
Just wanted to drop by and say thanks so much @roughike for all the great work here!

roughike commented:
@mit-mit And this is my way of saying thank you for everything that you guys have done for me :)

roughike commented:
Hey! Just a friendly Monday evening ping.

[The breaking change announcement on the flutter-dev mailing group](https://groups.google.com/forum/#!topic/flutter-dev/YE5tM2Y0BP8) has been live for 4 days now. So far, it hasn't gotten any comments. Being a noob to mailing groups, I accidentally posted it twice. Are we going to wait for a full week or how are things looking from that end? :)

Also, AFAIK, I've addressed all feedback and change requests. Please let me know if I missed something or if there's any additional work needed to be done.

goderbauer commented:
@roughike Sorry for the delay. I was OOO for the last few days. Once I am all caught up I will take another looks at this.

goderbauer commented:
@fkorotkov (you added the Cirrus CI to this repository): Any idea why Cirrus is consistently failing on this PR (even after multiple re-runs) without a clear exception whereas travis is just running fine?

goderbauer commented on packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/FileUtils.java:
```diff
@@ -0,0 +1,130 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
This is now missing the following note that @Hixie wanted included below the license. 

`
This file was modified by the Flutter authors from the following original file: https://raw.githubusercontent.com/iPaulPro/aFileChooser/master/aFileChooser/src/com/ipaulpro/afilechooser/utils/FileUtils.java
`

fkorotkov commented:
@goderbauer I think the issue is that this branch is pretty out dated. Cirrus CI does a single branch clone with only 50 commits in history. This seems to break logic in `script/incremental_build.sh` where we try to find a base sha for the branch.

I'll look into fixing `incremental_build.sh`, for now, @roughike could you please rebase your PR?

fkorotkov commented:
Fixed the `incremental_build.sh` issue in #444. Sorry for the inconvenience. 😅

roughike commented:
@fkorotkov Do you still need me to rebase the PR? :) I've never done that before and I was just figuring it out.

fkorotkov commented:
@roughike haven't you done so? 😅I see a merge commit b4c8b1d. In any case status checks ^^ are green so seems you did it and good to go 👌

roughike commented:
@fkorotkov Oh I see, I thought you meant something like this: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request. I thought it was going to be a long night for me. 👹

Hixie commented:
@fkorotkov fwiw, flutter needs the complete history to figure out its version number; not sure if that is relevant to your point or not. (Well, specifically, it needs the most recent tag and all the commits since then.)

fkorotkov commented:
@Hixie it might be an issue if/when we'll try to enable Cirrus on the main `flutter/flutter` repo. Here, Cirrus CI does a shallow clone only of the repo it's running a build for, e.g. `flutter/plugins`.

`cirrusci/flutter:latest` container on the other hand has full flutter repo cloned, so we are ok there 👌

goderbauer commented on packages/image_picker/lib/image_picker.dart:
```diff
@@ -6,47 +6,31 @@ import 'dart:async';
 import 'dart:io';
 
 import 'package:flutter/services.dart';
+import 'package:meta/meta.dart';
```
I was just trying to publish the new version and pub is complaining that package:meta is not a dependency declared in the pubspec.yaml. If you have a change, can you send a follow-up PR that ads the dependency and I'll then publish the new version.

roughike replied:
Weird, shouldn't it automatically come with Flutter? I've never had to define package:meta as a dependency in pubspec.yaml if I was already depending on Flutter.

Hixie replied:
import flutter foundation for meta

roughike replied:
@Hixie Thanks!
